core/node/state_keeper/src/seal_criteria/mod.rs

```diff
@@ -33,6 +33,114 @@ use super::{
+        Halt::FromIsNotAnAccount => "FromIsNotAnAccount",
+        Halt::InnerTxError => "InnerTxError",
+        Halt::Unknown(_) => "Unknown",
+        Halt::UnexpectedVMBehavior(_) => "UnexpectedVMBehavior",
+        Halt::BootloaderOutOfGas => "BootloaderOutOfGas",
+        Halt::ValidationOutOfGas => "ValidationOutOfGas",
+        Halt::TooBigGasLimit => "TooBigGasLimit",
+        Halt::NotEnoughGasProvided => "NotEnoughGasProvided",
+        Halt::MissingInvocationLimitReached => "MissingInvocationLimitReached",
+        Halt::FailedToSetL2Block(_) => "FailedToSetL2Block",
+        Halt::FailedToAppendTransactionToL2Block(_) => "FailedToAppendTransactionToL2Block",
+        Halt::VMPanic => "VMPanic",
+        Halt::TracerCustom(_) => "TracerCustom",
+        Halt::FailedToPublishCompressedBytecodes => "FailedToPublishCompressedBytecodes",
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ErrorMessage {
+    TxEncodingSize,
+    LargeEncodingSize,
+    PubdataLimit,
+    ProofWillFail,
+    TooMuchGas,
+    OutOfGasForBatchTip,
+    BootloaderOutOfGas,
+    NotEnoughGasProvided,
+}
+
+impl ErrorMessage {
+    pub fn as_static_str(&self) -> &'static str {
+        match &self {
+            ErrorMessage::TxEncodingSize => {
+                "Transaction cannot be included due to large encoding size"
+            }
+            ErrorMessage::LargeEncodingSize => {
+                "Transaction cannot be included due to large encoding size"
+            }
+            ErrorMessage::PubdataLimit => "Transaction cannot be sent to L1 due to pubdata limits",
+            ErrorMessage::ProofWillFail => "ZK proof cannot be generated for a transaction",
+            ErrorMessage::TooMuchGas => "Transaction requires too much gas",
+            ErrorMessage::OutOfGasForBatchTip => "Not enough gas for batch tip",
+            ErrorMessage::BootloaderOutOfGas => "Bootloader ran out of gas",
+            ErrorMessage::NotEnoughGasProvided => "Not enough gas provided",
+        }
+    }
+}
+
+fn error_message_as_metric_label(error_message: &ErrorMessage) -> &'static str {
+    match error_message {
+        ErrorMessage::TxEncodingSize => "TxEncodingSize",
+        ErrorMessage::LargeEncodingSize => "LargeEncodingSize",
+        ErrorMessage::PubdataLimit => "PubdataLimit",
+        ErrorMessage::ProofWillFail => "ProofWillFail",
+        ErrorMessage::TooMuchGas => "TooMuchGas",
+        ErrorMessage::OutOfGasForBatchTip => "OutOfGasForBatchTip",
+        ErrorMessage::BootloaderOutOfGas => "BootloaderOutOfGas",
+        ErrorMessage::NotEnoughGasProvided => "NotEnoughGasProvided",
+    }
+}
+/// Represents the reason variants for why a transaction was considered unexecutable.
+///
+/// This enum can capture either a hardcoded textual message or a more structured
+/// reason encapsulated in a `Halt` type.
+#[derive(Debug, Clone, PartialEq)]
+pub enum UnexecutableReason {
+    Text(ErrorMessage),
```

Comment:
TBH, I thought that variants you've defined in `ErrorMessage` would be included directly into `UnexecutableReason`:

```rust
pub enum UnexecutableReason {
    Halt(Halt),
    TxEncodingSize,
    LargeEncodingSize,
    PubdataLimit,
    ProofWillFail,
    TooMuchGas,
    OutOfGasForBatchTip,
    BootloaderOutOfGas,
    NotEnoughGasProvided,
}
```

Would it not work like that? This looks simpler; you don't require an additional enum and some of conversions.

Reply:
>This looks simpler

I do not feel so. There are two categories of UnexecutableReason: halt and hardcoded string. 
I think it is logical to single them out.
@perekopskiy?

Reply:
both options seem fine to me

Reply:
> There are two categories of UnexecutableReason: halt and hardcoded string

That's retroactive reasoning which I can't agree with. This is how it was implemented previously, yes, but this doesn't mean that this is the best approach.